lighthouse-core/config/budget.js

```diff
@@ -96,6 +98,97 @@ class Budget {
     };
   }
 
+  /**
+   * Validates that path is properly formed. Verifies the quantity and location
+   * of the two robot.txt regex characters: $, *
+   * @param {unknown} val
+   * @return {string}
+   */
+  static validatePath(val) {
+    /*
+     * If no path is specified the budget is assumed to apply to all pages. This makes
+     * budget.json backwards compatible with earlier API versions that did not include path.
+     */
+
+    if ((val === undefined) || (val === null)) {
+      return '/';
+    }
+    const path = /** @type {string} */ (val);
+    const hasLeadingSlash = path.startsWith('/');
+    const validWildcardQuantity = ((path.match(/\*/g) || []).length <= 1);
+    const validDollarSignQuantity = ((path.match(/\$/g) || []).length <= 1);
+    const validDollarSignPlacement = !path.includes('$') || path.endsWith('$');
+
+    const isValid = hasLeadingSlash && validWildcardQuantity
+      && validDollarSignQuantity && validDollarSignPlacement;
+
+    if (!isValid) {
+      throw new Error(`Invalid path ${path}. ` +
+        `'Path' should be specified using the 'robots.txt' format.\n` +
+        `Learn more about the 'robots.txt' format here:\n` +
+        `https://developers.google.com/search/reference/robots_txt#url-matching-based-on-path-values`);
+    }
+    return path;
+  }
+
+  /**
+   * Determines whether a URL matches against a robots.txt-style "path".
+   * @param {string} url
+   * @param {string} pattern
+   * @return {boolean}
+   */
+  static urlMatchesPattern(url, pattern) {
```

Comment:
LOVE the documentation and style of this function here :D

Reply:
> LOVE the documentation and style of this function here :D

agreed, this is great